### PR TITLE
Add GGJT loader

### PR DESCRIPTION
**Pull request by iacore**
Related to #62 #93 

single-file model format magic=ggjt

### CODE REVIEW

**Comment by philpax** [comment]
Very nice! Glad someone took this on - we were discussing this on the Discord but decided to wait until upstream figured out what they wanted to do.

There's overlap with #85, so the merge in the future might be tricky. Just a heads-up.

**Comment by iacore** [comment]
I ported the code from C++.

I'm sure the reading part is correct. Maybe Tensor setup has changed? Or ggml has also changed?

**Comment by philpax** [comment]
I'd take the same model in ggmj and ggjt format and compare the loaded tensors. I'm guessing it's probably a misalignment or something similar.

**Comment by iacore** [comment]
Is it normal to warn for bad tokens? The bad tokens are single-byte string >128. That's invalid UTF-8, but the tokenizer don't care?

**Comment by philpax** [comment]
Yes, we're discussing that over at #11. The short explanation is that the invalid tokens are not valid UTF8 by themselves, but are composed to form valid UTF8. We're figuring out what the actual solution should be.

**Comment by KerfuffleV2** [comment]
Does it work with the old formats but just gets garbage when using a GGJT model or just always garbage?

**Comment by iacore** [comment]
I haven't change the behavior of old formats. only GGJT produce garbage

**Comment by KerfuffleV2** [comment]
## `ggml`

Fine.

## `ggmf`

```plaintext
thread 'main' panicked at 'Could not load model: TensorWrongSize { tensor_name: "tok_embeddings.weight", path: "blah.bin" }', llama-cli/src/main.rs:206:6
```

(I don't know if it's actually supposed to work here or not.)

## `ggjt`

Garbage.

**Comment by KerfuffleV2** [comment]
Was that last force push just rebasing on the current version or does it involve changes to the loading that may fix stuff that previously didn't work?

**Comment by iacore** [comment]
> Was that last force push just rebasing on the current version or does it involve changes to the loading that may fix stuff that previously didn't work?

just rebase

**Comment by KerfuffleV2** [comment]
I think this one is GGMF: https://huggingface.co/Sosaka/Alpaca-native-4bit-ggml/

(Don't know if it's a problem to post something like that here, if so let me know and I'll edit it out after iacore sees it.)

**Comment by iacore** [comment]
+1

**Comment by KerfuffleV2** [comment]
Sorry, my mistake. Unfortunately, I don't really have a reasonable way to share huge files.

I've been trying to figure out what the issue with GGJT is. One thing I can say is I don't think your logic for finding the tensors/lengths/offsets has a problem.

I added some `printf`s to `llama.cpp` and the corresponding ones to `llama-rs`.

```plaintext
Load: tok_embeddings.weight, offset=432672, size=102403200
Load: norm.weight, offset=102835904, size=20480
Load: output.weight, offset=102856448, size=102403200
[...]
Load: layers.39.feed_forward.w2.weight, offset=8048282880, size=44236800
Load: layers.39.feed_forward.w3.weight, offset=8092519744, size=44236800
Load: layers.39.ffn_norm.weight, offset=8136756608, size=20480
```

Absolutely no difference in output between the C and Rust versions.

**Comment by iacore** [comment]
still no clue

**Comment by KerfuffleV2** [comment]
Okay, so I got it actually running inference on a GGJT model. However, what I had to do makes the mmap part pointless.

I believe the problem has something to do with how you can set a context parameters to `no_alloc`. In the llama.cpp change that added the format, they set `no_alloc` to true for the main context and then reduce the context size a lot so that GGML doesn't allocate memory for the actual tensors.

However, we're still doing the old context size calculation. I tried making the `Context::init` function take a bool for no_alloc and set it, but just got a segfault immediately.

Anyway, in `loader.rs:load_weights_ggjt` just change:

```rust
tensor.set_data(ptr as *mut std::ffi::c_void);
```

to

```rust
ptr.copy_to_nonoverlapping(tensor.data() as *mut u8, tensor.nbytes());
```

With that, it runs just fine on the GGJT model. Loading speed seems normal compared to the current version.

Obviously it's silly to `mmap` and just copy the data immediately. What I'd recommend is just ditching `mmap` right now and simply reading into the tensor data instead.

Then later on it will be possible to add `mmap` support as a separate thing which could work in a general way for the other formats too.

By the way, you probably need to run clippy on your changes. It's very unhappy right now!

**Comment by philpax** [comment]
Yeah, I'd be happy with not supporting `mmap` right now. We can figure out what that's meant to look like once we have support for all the model types working.

**Comment by KerfuffleV2** [comment]
The approach I'd go for if I was writing it would be to just have a general type that can describe the types of tensor and where they are, dimensions, etc. Something similar to this: https://github.com/KerfuffleV2/smolrsrwkv/blob/182cd3205b7a7c95571a09bcfbb954b0041e4f90/smolrwkv/src/loader.rs#L15

Then the specific file format code can just scan through the file for metadata, build a structure with a list of those things or whatever. Then there can be generic loader code that just loads tensors based on that: it could use reading files, mmap, whatever.

I think that approach would make dealing with different file formats _much_ easier.

**Comment by philpax** [comment]
I'd be OK with that. It'd also help with #117 / #84.

**Comment by iacore** [comment]
ship it

**Comment by KerfuffleV2** [comment]
@iacore I was experimenting with trying to clean it up also. It does seem like reading is way slower than the copying from `mmap` approach. I don't know why.

My change looks like this:

```rust
        let offset_curr = reader.stream_position()?;
        let offset_aligned: u64 = (offset_curr + 31) & !31;

        reader.seek_relative((offset_aligned - offset_curr) as i64)?;
        let td =
            unsafe { std::slice::from_raw_parts_mut(tensor.data() as *mut u8, tensor.nbytes()) };

        reader.read_exact(td)?;
        total_loaded_bytes += tensor.nbytes();
```

So, uhh... I guess maybe just keep mmap?

**Comment by iacore** [comment]
for me, seq. read is faster than mmap

**Comment by KerfuffleV2** [comment]
Are you saying you tried the `copy_to_nonoverlapping` version and it was slower than the current version that changed to `BufReader`?

**Comment by iacore** [comment]
> Are you saying you tried the `copy_to_nonoverlapping` version and it was slower than the current version that changed to `BufReader`?

yes

**Comment by KerfuffleV2** [comment]
Ahh, why can't life be simple for once?

What OS are you using, out of curiosity?

**Comment by iacore** [comment]
Linux.

I think this branch is done. I'll not touch it again.

**Comment by KerfuffleV2** [comment]
I'm on Linux as well.

Sorry for the confusion. I've been trying both versions and I don't get consistent results. No sure what's going on, but I don't think there's a problem with the current approach.

> I'll not touch it again.

How can you say that when Clippy is still sad?

You can probably just nuke the `set_data` method, it doesn't seem clear that there's even a way to successfully use it at the moment.

**Comment by iacore** [comment]
> You can probably just nuke the set_data method, it doesn't seem clear that there's even a way to successfully use it at the moment.

it probably will be useful in the future?

**Comment by KerfuffleV2** [comment]
> it probably will be useful in the future?

We'd have to figure out how to actually use it in the future though. I don't think it currently can work at all, since you aren't even able to turn of `need_alloc` when creating a context so memory for tensors will always get allocated no matter what.

You'd think it would still be possible to set the tensor to point at a different chunk of memory but that didn't actually work: otherwise your first approach would have had no issues.

So my line of thinking is, it's a only a couple of lines of code to wrap a `ggml` function so it wouldn't be hard to add back in later on if it was actually needed/could be used but it's currently non-functional so it may as well be removed.

Just to be clear, this is just the opinion of some other random person on the internet. So take that for it's worth, I have no authority.

**Comment by jon-chuang** [comment]
> I think this branch is done. I'll not touch it again.

So, we're not gonna get the new mmap format cuz poor perf?

**Comment by jon-chuang** [comment]
I also got this working on new ggjt file format.

**Comment by jon-chuang** [comment]
Btw, from https://justine.lol/mmap/

> Remember that progress bar which made you wait for weights to load each time you ran the command? We got rid of that. Linux users should expect a 100x improvement in load time. Windows and MacOS users should expect a 10x improvement. What this means is that tokens will start being produced effectively instantaneously when you run LLaMA, almost providing a similar UX to ChatGPT on the shell. It's important to note these improvements are due to an amortized cost. The first time you load a model after rebooting your computer, it's still going to go slow, because it has to load the weights from disk. However each time it's loaded afterwards, it should be fast (at least until memory pressure causes your file cache to be evicted).

The speedup is meant to be for subsequent loads. Did you guys check that?

**Comment by jon-chuang** [comment]
Another thing to check: multiple processes utilizing the same mmaped file:
> More Processes
You can now run multiple LLaMA processes simultaneously on your computer. Here's a [video of Georgi having a conversation with four chatbots](https://twitter.com/ggerganov/status/1642115206544343040) powered by four independent llama.cpp processes running on the same Mac. So llama.cpp is not only going to be a better friend to you, it can also serve as your artificial circle of friends too. The trick that makes it possible is mmap() lets us map the read-only weights using MAP_SHARED, which is the same technique that's traditionally been used for loading executable software. So we figured, why aren't we using it to load neural network software too? Now we can.

File: llama-rs/src/lib.rs [hunk 1]
```diff
@@ -828,243 +841,28 @@ impl Model {
...
+                drop(reader);
+                load_weights_ggmf_or_unversioned(
+                    file_offset,
+                    main_path,
+                    load_progress_callback,
+                    &model,
+                )?
+            }
+            ModelVersion::GGJT => {
+                let mmap = unsafe { Mmap::map(&file)? };
```
Comment (jon-chuang) [inline llama-rs/src/lib.rs:861]: Can add advise
https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#method.advise
here with
https://docs.rs/memmap2/latest/memmap2/enum.Advice.html#variant.Sequential

**Comment by iacore** [comment]
Superceded by #125

**Pull request closed without merging**